llama_index/response_synthesizers/base.py

```diff
@@ -0,0 +1,173 @@
+
+    def __init__(
+        self,
+        service_context: Optional[ServiceContext] = None,
+        streaming: bool = False,
+    ) -> None:
+        """Init params."""
+        self._service_context = service_context or ServiceContext.from_defaults()
+        self._callback_manager = self._service_context.callback_manager
+        self._streaming = streaming
+
+    @property
+    def service_context(self) -> ServiceContext:
+        return self._service_context
+
+    @abstractmethod
+    def get_response(
+        self,
+        query_str: str,
+        text_chunks: Sequence[str],
+        **response_kwargs: Any,
+    ) -> RESPONSE_TEXT_TYPE:
+        """Get response."""
+        ...
+
+    @abstractmethod
+    async def aget_response(
+        self,
+        query_str: str,
+        text_chunks: Sequence[str],
+        **response_kwargs: Any,
+    ) -> RESPONSE_TEXT_TYPE:
+        """Get response."""
+        ...
+
+    def _log_prompt_and_response(
+        self,
+        formatted_prompt: str,
+        response: RESPONSE_TEXT_TYPE,
+        log_prefix: str = "",
+    ) -> None:
+        """Log prompt and response from LLM."""
+        logger.debug(f"> {log_prefix} prompt template: {formatted_prompt}")
+        self._service_context.llama_logger.add_log(
+            {"formatted_prompt_template": formatted_prompt}
+        )
+        logger.debug(f"> {log_prefix} response: {response}")
+        self._service_context.llama_logger.add_log(
+            {f"{log_prefix.lower()}_response": response or "Empty Response"}
+        )
+
+    def _get_metadata_for_response(
+        self,
+        nodes: List[BaseNode],
+    ) -> Optional[Dict[str, Any]]:
+        """Get metadata for response."""
+        return {node.node_id: node.metadata for node in nodes}
+
+    def _prepare_response_output(
+        self,
+        response_str: Optional[RESPONSE_TEXT_TYPE],
+        source_nodes: List[NodeWithScore],
+    ) -> RESPONSE_TYPE:
+        """Prepare response object from response string."""
+        response_metadata = self._get_metadata_for_response(
+            [node_with_score.node for node_with_score in source_nodes]
+        )
+
+        if response_str is None or isinstance(response_str, str):
+            return Response(
+                response_str,
+                source_nodes=source_nodes,
+                metadata=response_metadata,
+            )
+        elif response_str is None or isinstance(response_str, Generator):
+            return StreamingResponse(
+                response_str,
+                source_nodes=source_nodes,
+                metadata=response_metadata,
+            )
+        else:
+            raise ValueError(
+                f"Response must be a string or a generator. Found {type(response_str)}"
+            )
+
+    def synthesize(
+        self,
+        query: QueryTextType,
+        nodes: List[NodeWithScore],
+        additional_source_nodes: Optional[Sequence[NodeWithScore]] = None,
```

Comment:
(should be for a later PR) was chatting with @sourabhdesai a bit on this last week. we should think of a better way of returning source nodes instead of just flat list, especially when responses become hierarchical (e.g. in our graph query engine, the recursive retriever query engine, sub question query engine).

Reply:
True. At that point, there should be some mechanism for the query engine itself to handle the response source structure? 

Well, would take some planning. Sounds like a PR that needs to happen before Wednesday though to meet the feature complete deadline 😉